// @stackworx/react-hook-form-mui-x-date-pickers/dist/DateTimePicker.js
import { jsx } from "react/jsx-runtime";
import { useController } from "react-hook-form";
import { DateTimePicker as DateTimePicker$1 } from "@mui/x-date-pickers/DateTimePicker";
import { useUtils, useLocalizationContext, validateDateTime } from "@mui/x-date-pickers/internals";
import { mapDateTimePickerValidationErrorMessage } from "./utils/DatePickerValidationErrorsToMessage.js";
function DateTimePicker({
  name,
  rules,
  control,
  ...props
}) {
  var _a, _b;
  const { slotProps, ...otherPickerProps } = props;
  const { getTimezone } = useUtils();
  const adapter = useLocalizationContext();
  const {
    field: { onChange, value, ref, onBlur },
    fieldState
  } = useController({
    name,
    control,
    rules: {
      ...rules,
      validate: {
        ...rules == null ? void 0 : rules.validate,
        internalMuiError: () => {
          const muiValidationError = validateDateTime({
            value,
            props: {
              ...props,
              disableFuture: !!props.disableFuture,
              disablePast: !!props.disablePast,
              minDate: props.minDateTime ?? props.minDate,
              maxDate: props.maxDateTime ?? props.maxDate,
              minTime: props.minDateTime ?? props.minTime,
              maxTime: props.maxDateTime ?? props.maxTime,
              timezone: getTimezone(value)
            },
            adapter
          });
          return mapDateTimePickerValidationErrorMessage(
            muiValidationError,
            props
          ) ?? true;
        }
      }
    }
  });
  return /* @__PURE__ */ jsx(
    DateTimePicker$1,
    {
      onChange,
      value,
      ref,
      slotProps: {
        ...slotProps,
        textField: {
          ...slotProps == null ? void 0 : slotProps.textField,
          inputRef: ref,
          error: !!fieldState.error,
          onBlur,
          helperText: ((_a = fieldState.error) == null ? void 0 : _a.message) ?? ((_b = slotProps == null ? void 0 : slotProps.textField) == null ? void 0 : _b.helperText)
        }
      },
      ...otherPickerProps
    }
  );
}
export {
  DateTimePicker
};
