// @stackworx/react-hook-form-mui/dist/CheckboxGroup.js
import { jsx } from "react/jsx-runtime";
import { useController } from "react-hook-form";
import MuiCheckbox from "@mui/material/Checkbox";
function CheckboxGroup({
  control,
  name,
  rules,
  value,
  ...props
}) {
  const {
    field,
    fieldState: { error }
  } = useController({
    name,
    control,
    rules
  });
  const { onChange, onBlur, ref } = field;
  return /* @__PURE__ */ jsx(
    MuiCheckbox,
    {
      sx: {
        ...error && {
          color: "error.main",
          "&.Mui-checked": {
            color: "error.main"
          }
        }
      },
      ...props,
      checked: field.value.includes(value),
      inputRef: ref,
      onChange: (_event, checked) => {
        if (checked) {
          onChange([...field.value, value]);
        } else {
          onChange(
            // @ts-expect-error must be array
            field.value.filter((v) => v !== value)
          );
        }
      },
      onBlur,
      value,
      name
    }
  );
}
CheckboxGroup.displayName = "MuiReactHookFormCheckboxGroup";
export {
  CheckboxGroup
};
