// react-relay/lib/legacy.js
'use strict';

var ReactRelayContext = require('./ReactRelayContext');
var ReactRelayFragmentContainer = require('./ReactRelayFragmentContainer');
var ReactRelayLocalQueryRenderer = require('./ReactRelayLocalQueryRenderer');
var ReactRelayPaginationContainer = require('./ReactRelayPaginationContainer');
var ReactRelayQueryRenderer = require('./ReactRelayQueryRenderer');
var ReactRelayRefetchContainer = require('./ReactRelayRefetchContainer');
var RelayRuntime = require('relay-runtime');
module.exports = {
  ConnectionHandler: RelayRuntime.ConnectionHandler,
  QueryRenderer: ReactRelayQueryRenderer,
  LocalQueryRenderer: ReactRelayLocalQueryRenderer,
  MutationTypes: RelayRuntime.MutationTypes,
  RangeOperations: RelayRuntime.RangeOperations,
  ReactRelayContext: ReactRelayContext,
  applyOptimisticMutation: RelayRuntime.applyOptimisticMutation,
  commitLocalUpdate: RelayRuntime.commitLocalUpdate,
  commitMutation: RelayRuntime.commitMutation,
  createFragmentContainer: ReactRelayFragmentContainer.createContainer,
  createPaginationContainer: ReactRelayPaginationContainer.createContainer,
  createRefetchContainer: ReactRelayRefetchContainer.createContainer,
  fetchQuery_DEPRECATED: RelayRuntime.fetchQuery_DEPRECATED,
  graphql: RelayRuntime.graphql,
  readInlineData: RelayRuntime.readInlineData,
  requestSubscription: RelayRuntime.requestSubscription
};
